rcli/typing.py

```diff
@@ -0,0 +1,671 @@
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound), keyfunc_name)
+        else:
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound))
+        return _BoundedSubclass
+
+    def _get_args(cls, args):
+        # type: (tuple) -> Tuple[Any, slice, Callable]
+        """Return the parameters necessary to check type boundaries.
+
+        Args:
+            args: A tuple with two or three elements: a type, a slice
+                representing the minimum and maximum lengths allowed for values
+                of that type and, optionally, a function to use on values
+                before comparing against the bounds.
+
+        Returns:
+            A tuple with three elements: a type, a slice, and a function to
+            apply to objects of the given type. If no function was specified,
+            it returns the identity function.
+        """
+        if not isinstance(args, tuple):
+            raise TypeError(
+                '{}[...] takes two or three arguments.'.format(cls.__name__))
+        elif len(args) == 2:
+            type_, bound = args
+            keyfunc = cls._identity
+        elif len(args) == 3:
+            type_, bound, keyfunc = args
+        else:
+            raise TypeError(
+                'Too many parameters given to {}[...]'.format(cls.__name__))
+        if not isinstance(bound, slice):
+            bound = slice(bound)
+        if isinstance(type_, six.string_types):
+            # pragma pylint: disable=protected-access
+            type_ = _ForwardRef(type_)._eval_type(globals(), globals())
+            # pragma pylint: enable=protected-access
+        return type_, bound, keyfunc
+
+    @staticmethod
+    def _get_bound_repr(bound):
+        # type: (slice) -> str
+        """Return a string representation of a boundary slice.
+
+        Args:
+            bound: A slice object.
+
+        Returns:
+            A string representing the slice.
+        """
+        if bound.start and not bound.stop:
+            return '{}:'.format(bound.start)
+        if bound.stop and not bound.start:
+            return ':{}'.format(bound.stop)
+        return '{}:{}'.format(bound.start, bound.stop)
+
+    @staticmethod
+    def _identity(obj):
+        # type: (_T) -> _T
+        """Return the given object.
+
+        Args:
+            obj: An object.
+
+        Returns:
+            The given object.
+        """
+        return obj
+
+    _default = _identity  # type: Callable[[Any], Any]
+
+    @staticmethod
+    def _get_fullname(obj):
+        # type: (Any) -> str
+        """Get the full name of an object including the module.
+
+        Args:
+            obj: An object.
+
+        Returns:
+            The full name of the object.
```

Comment:
It turns out that instances of User classes don't have `__name__`'s.

I think this means that the right thing to do is to change the docstring  of this method to claim that the full classname is extracted.